### PR TITLE
fix knock retard table init

### LINKED ISSUE #246: Knock retard not activating (20230516 release)
Based on the testing we did at silverstone it seems the knock retard is not activating, knock count increases, and the volume on the sensors passes the threshold but no timing is removed. 

Over the day we tried several combinations of aggression (5 and 10%), threshold (lowered it by 2dBv) and max allowed retard (tried several tables then flat mapped) with no change in result. 

Logs and tune attached.  
MLV knock polynomial is " (-0.000000814*[RPM]*[RPM]) + (0.00986*[RPM]) - 37 + 2 "

[re_17.zip](https://github.com/FOME-Tech/fome-fw/files/12640948/re_17.zip)

### PR DESCRIPTION
**Pull request by nmschulte**
column and row definition were swapped w/re: value lookup

fixes #246, #417

----

Andrey was right: https://github.com/FOME-Tech/fome-fw/issues/417#issuecomment-2087816779
> is this https://github.com/rusefi/rusefi/issues/6370?

Comparison to boost; RPM is columns:
2024-05-02_02h00m55s_screenshot

boost:
```cxx
// init
boostMapOpen.init(config->boostTableOpenLoop, config->boostTpsBins, config->boostRpmBins);
boostMapClosed.init(config->boostTableClosedLoop, config->boostTpsBins, config->boostRpmBins);

// getValue
float openLoop = luaOpenLoopAdd + m_openLoopMap->getValue(rpm, driverIntent.Value);
float target = m_closedLoopTargetMap->getValue(rpm, driverIntent.Value);
```

knock:
```cxx
// init
m_maxRetardTable.init(config->maxKnockRetardTable, config->maxKnockRetardRpmBins, config->maxKnockRetardLoadBins);

// getValue
return m_maxRetardTable.getValue(Sensor::getOrZero(SensorType::Rpm), getIgnitionLoad());
```

### CODE REVIEW

**Comment by rusefillc** [comment]
While you are fixing this defect root cause is order of arguments on the method. See rusefi changes flipping it.

**Comment by nmschulte** [comment]
> fixes #246, #417

I haven't confirmed this; this statement might be an overreach.

I'm also not sure this mistake explains how `m_maximumRetard` would be zero if the table is entirely non-zero.

----

> While you are fixing this defect root cause is order of arguments on the method. See rusefi changes flipping it.

I'll check that out soon; had looked at it just a little before.

As well, the type system should be able to save us from mistakes like this somehow.  "traits" naively come to mind but I'm admittedly unfamiliar; the standard kit (pre-traits) should be able to do it though too if I'm not mistaken.

**Comment by nmschulte** [comment]
@rk-iv has effected this change by altering the TS `.ini`, like so; remapping the bins and the table axes so the `m_maxRetardTable->init` logic does what it should and TS editor can be used as intended w/o brain games:

```diff
 maxKnockRetardTable = array, U08, 19556, [6x6], "deg", 0.25, 0, 0, 30, 2
-maxKnockRetardLoadBins = array, U08, 19592, [6], "%", 1, 0, 0, 250, 0
-maxKnockRetardRpmBins = array, U08, 19598, [6], "RPM", 100.0, 0, 0, 25000, 0
+maxKnockRetardRpmBins = array, U08, 19592, [6], "%", 1, 0, 0, 250, 0
+maxKnockRetardLoadBins = array, U08, 19598, [6], "RPM", 100.0, 0, 0, 25000, 0
```

```diff
     table = maxKnockRetardTbl, maxKnockRetardMap, "Max knock retard",    1
-        xBins        = maxKnockRetardRpmBins, RPMValue
-        yBins        = maxKnockRetardLoadBins, ignitionLoad
+        xBins        = maxKnockRetardLoadBins, RPMValue
+        yBins        = maxKnockRetardRpmBins, ignitionLoad
         zBins        = maxKnockRetardTable
         gridOrient  = 250,    0, 340 ; Space 123 rotation of grid in degrees.
```

It doesn't fully solve the problem; the CLFC off->on dance is still required to achieve desired effect.

**Comment by nmschulte** [comment]
I found that `KnockController` module had no "initialize" method, and so `m_maximumRetard` was not being initialized on ECU reset.  Added an implementation similar to that of `BoostController`; probably not perfect but should resolve the issue.

File: firmware/controllers/engine_cycle/knock_controller.cpp [hunk 1]
```diff
@@ -8,10 +8,14 @@
 #include "pch.h"
 #include "knock_logic.h"
 
+void KnockController::init(engine_configuration_s const * const previousConfig) {
+	m_maxRetardTable.init(config->maxKnockRetardTable, config->maxKnockRetardLoadBins, config->maxKnockRetardRpmBins);
```
Comment (nmschulte) [inline firmware/controllers/engine_cycle/knock_controller.cpp:12]: note this flips the load/rpm bins to match the current implementation

**Comment by rusefillc** [comment]
> I found that KnockController module had no "initialize" method, and so m_maximumRetard was not being initialized on ECU reset.

if that's the case, and it seems to be the case, I assume the most important next step is to make sure such a defect never occurs again? https://github.com/rusefi/rusefi/issues/6461

File: firmware/controllers/engine_cycle/knock_controller.cpp [hunk 2]
```diff
@@ -8,10 +8,14 @@
 #include "pch.h"
 #include "knock_logic.h"
 
+void KnockController::init(engine_configuration_s const * const previousConfig) {
```
Comment (rusefillc) [inline firmware/controllers/engine_cycle/knock_controller.cpp:11]: what is the point of 'previousConfig' here?

Comment (nmschulte) [reply firmware/controllers/engine_cycle/knock_controller.cpp]: you are correct, I was moving too fast, didn't realize `config` was "global" and `previousConfig` is not what `onConfigurationChange` was/is using

**Comment by nmschulte** [comment]
> I assume the most important next step is to make sure such a defect never occurs again?

I agree, see also #422 (and #421).

File: firmware/controllers/engine_cycle/knock_controller.cpp (hunk 1 above)
Comment (mck1117) [reply firmware/controllers/engine_cycle/knock_controller.cpp:12]: not the minimal diff to fix the bug, sigh

**Pull request merged**